Update preprocess.py for normalization

ZechangSun opened the pull request:
Dear X, this code is about the normalization, hope for your advice. Thanks!

ZechangSun committed:
Update preprocess.py for normalization

ZechangSun committed:
Update DesiMock.py

profxj commented:
Ok, I will review this once the PR #33 is done.
Thanks.

ZechangSun committed:
Add files via upload

profxj commented on dla_cnn/desi/DesiMock.py:
```diff
@@ -88,6 +95,18 @@ def get_data(start_point = 0, end_point = self.data_size):
             sightline.loglam = np.log10(self.wavelength[start_point:end_point])
+
+            wavelength = self.wavelength[start_point:end_point]
+            dlambda = wavelength-np.roll(wavelength,1)
```
Please explain what these lines of code are for.
I'll note there will always be at least 1 negative value because `np.roll` wraps the array around.

ZechangSun replied:
Emmm, I just chose the whole array at this place. And I did the slicing when I really needed it like this ` test = (dlambda>0)[1:]` . Sorry for confusing you.

profxj commented on dla_cnn/desi/DesiMock.py:
```diff
@@ -88,6 +95,18 @@ def get_data(start_point = 0, end_point = self.data_size):
+            test = (dlambda>0)[1:]
+            indice = np.argwhere(~test)
+            
+            if indice.size:# the wavelength array may not be monotonic increasing, so we do a test here.
```
Why would it not be monotonically increasing?  
If it isn't this should mean there is a problem and 
you should `raise` an error.

ZechangSun replied:
I am not sure, only a very small part of the data have this problem. Maybe we can still use  these spectrum although they have this problem. I am not sure...

profxj replied:
Let's try to understand what this occurs.
Can you provide an example on Slack?

ZechangSun replied:
Sure~

profxj replied:
This was a bug.

profxj commented on dla_cnn/desi/preprocess.py:
```diff
@@ -92,3 +92,33 @@ def label_sightline(sightline, kernel, REST_RANGE, pos_sample_kernel_percent=0.3
+    sightline: the sightline after normalized
+    
+    '''
+    def _normalize(sightline,left,right):
```
please rename this method to not be identical to the parent method.
too confusing.

profxj replied:
And add a short doc string for it too.

profxj commented on dla_cnn/desi/preprocess.py:
```diff
@@ -92,3 +92,33 @@ def label_sightline(sightline, kernel, REST_RANGE, pos_sample_kernel_percent=0.3
+        test =(rest_wavelength>=left)&(rest_wavelength<=right)
+        sightline.flux = sightline.flux/np.median(sightline.flux[test])
+        
+    if camera == 'b':
```
This isn't going to work well in general.

Here are two different options to consider:

1.  Find which camera contains 1450A in the rest frame.  And use it to normalize all cameras.
2.  Find the camera with the most coverage of the Lya forest (1070-1170) and use that instead.

I somewhat prefer the 2nd option.

profxj commented:
This is a little more challenging then what we have tried so far.
In particular the quasars will have a wide range of redshifts.

I've suggested 2 options above.  Please explore these for a 
set of ~10 quasars with z=2-4.  Thanks.

ZechangSun committed:
Update preprocess.py

ZechangSun committed:
Update DesiMock.py

ZechangSun committed:
Update preprocess.py

ZechangSun committed:
Update DesiMock.py

ZechangSun committed:
Update DesiMock.py

ZechangSun committed:
Merge branch 'master' into normalize

ZechangSun committed:
Delete methods of normalization.ipynb

ZechangSun committed:
Add files via upload

ZechangSun committed:
Update defs.py

ZechangSun committed:
Delete methods of rebining &&determine the best v for each camera .ipynb

ZechangSun committed:
Add files via upload

ZechangSun commented:
I finally chose the 1st option since [1070,1170] is out of some spectrum's wavelength range. I have also squashed the bug and updated the previous work which may have problems because of that bug. I have updated a notebook which shows how to normalize the spectrum and 19 normalized spectrum redshift from 2 to 4. Hope for your advice. Thanks!

profxj commented:
Ok, a few things:

1. Which spectra do not cover [1070, 1170]?  Or better put, if they don't then we can't search for DLAs, right?
2.  I'd like to see the normalized data that cover the Lya forest.  This is what we will be analyzing.

Thanks!

profxj commented on dla_cnn/desi/DesiMock.py:
```diff
@@ -64,14 +65,25 @@ def read_fits_file(self, spec_path, truth_path, zbest_path):
+                rest_wavelength = self.wavelength/(data['z_qso']+1)
+                if rest_wavelength[0]<1450:
+                    test = (rest_wavelength>=1400)&(rest_wavelength<=1500)
+                    data['FLUX'] = data['FLUX']/np.abs((np.median(data['FLUX'][test])))
```
Why do you take `np.abs` of the median?  Surely it should never be negative.

profxj replied:
How about of instead of actually applying the normalization that you record it in `self.data` and we can choose to use it later when generating the `Sightline` object?

This would actually allow you to calculate both the normalization values we are considering.

profxj replied:
Also, we need to renormalize the error array by the same factor to maintain the signal/noise.

ZechangSun committed:
Update DesiMock.py

ZechangSun commented:
Thanks for your advice above. I discussed with @caiz12, and his advice is:
1.  If (1+z_qso)*1216-3000/c *1216 *(1+z_qso) is bluer than 3800A, the sightline is not needed, because of the low s/n in the part lower than 3800A;
2. when do normalizing, decide which is bluer, 3800A or rest-frame 1070A. If latter, use 3800A, that is due to the low s/n in the part lower than 3800A.
I have corrected the other aspects as you have suggested. Thanks!

ZechangSun committed:
Delete methods of normalization.ipynb

ZechangSun committed:
Add files via upload

profxj commented on dla_cnn/desi/DesiMock.py:
```diff
@@ -65,36 +70,32 @@ def read_fits_file(self, spec_path, truth_path, zbest_path, normalize = False):
-            assert err_spec == '', "Something wrong with spectrum below, please check them: \n"+err_spec
-            
+        
+        ndlas = []
```
I see no reason to remove the spectra here.  
At the least, make it optional and default to not doing it.

We will choose which quasars (and DLAs) to include
when building our training, test sets, etc.

ZechangSun replied:
Emmm, I remain those spectra, but some of them may not contain the Lymann forest part(3800/(1+z_qso)>1170, and in that case, the normalize method will assert error. So the users may need some test when using the normalize method. Maybe we can drop those spectrum with z<2.3, so all spectrum in DesiMock can be normalized.

profxj replied:
No, best that we error out if we try to normalize a spectrum that we shouldn't.

profxj commented on dla_cnn/desi/DesiMock.py:
```diff
@@ -65,36 +70,32 @@ def read_fits_file(self, spec_path, truth_path, zbest_path, normalize = False):
 
 
-    def get_sightline(self, id, camera = 'all', rebin = False):
+    def get_sightline(self, id, camera = 'all', rebin = False, normalize = False):
```
The parameter `normalize` is never used in this method.

Looks like a bug as you then use `_normalize` in the internal `def get_data()` method.

profxj commented on dla_cnn/desi/DesiMock.py:
```diff
@@ -109,7 +110,13 @@ def get_data(start_point = 0, end_point = self.data_size):
-
+            
+            if _normalize:
+                central = max(3800/(1+sightline.z_qso),1070)
```
If `central == 1150`  (as it will for some), then 
you will be measuring the flux from 1150-1250 
not 1150-1170.  This is a bug.

profxj replied:
Also, let's make normalize its own method in `preprocess.py`

That way we can even have multiple options.  And this is a 
preprocessing step anyhow.  Not I/O

ZechangSun replied:
Thanks for your advice.I have discussed with @caiz12 , and we chose to drop those spectra whose `central>1170`, so we use 1170 as the upper limit for the slice. That's how we deal with this bug.
And I have removed the `_nomalize` method to the `preprocess.py`, but I also remains the parameter `nomalize` in `get_sightline`, I think it may be convenient in practice.

profxj commented on dla_cnn/desi/DesiMock.py:
```diff
@@ -65,36 +70,32 @@ def read_fits_file(self, spec_path, truth_path, zbest_path, normalize = False):
                      'z' : Load up the wavelength and data for the z camera, 'all':  Load up the wavelength and data for all cameras.
-        rebin: bool, if True rebin the spectra to the best dlambda/lambda, default False.
+        rebin: bool, if True rebin the spectra to the best dlambda/lambda, default False,
+        normalize: bool, if True normalize the spectra, default False.
```
Add docs on the normalization procedure.

ZechangSun replied:
I have did that, thanks for remaining.

profxj commented:
Ok, this is almost good.  Address the bugs and
move the normalization code to its own method in `preprocess.py`.

We'll have to see how this works on low S/N data, but
hopefully it will be ok.  Thanks @ZechangSun

ZechangSun committed:
Update DesiMock.py

ZechangSun committed:
Update preprocess.py

ZechangSun committed:
Delete methods of normalization.ipynb

profxj committed:
jxp edits

profxj committed:
jxp edits

ZechangSun committed:
Update preprocess.py

ZechangSun committed:
Update Sightline.py

ZechangSun committed:
Update DesiMock.py

ZechangSun committed:
Update preprocess.py

ZechangSun commented:
Dear X,
      I noticed that you have reminded me that some comments needed to be added to make my code more readable, I have updated my code and added those comments. Thanks for your advice. Besides, I have also noticed that you have changed the code slightly for its readability, I really appreciate your advice, my previous work is confusing sometimes, and I have learnt from your suggestions and will make sure its readability next time. I have updated the other parts of the code to be consistent with the changes you have made. 
   Wish your good healthy and thanks for your help!

ZechangSun committed:
Add files via upload

profxj commented:
Great!

I am merging now.

profxj reviewed (approved):
approving

profxj commented:
merging.
